zap/src/main/java/org/parosproxy/paros/core/scanner/AbstractAppVariantPlugin.java

```diff
@@ -0,0 +1,76 @@
+package org.parosproxy.paros.core.scanner;
+
+import java.util.List;
+import org.apache.log4j.Logger;
+import org.parosproxy.paros.Constant;
+import org.parosproxy.paros.model.Model;
+import org.parosproxy.paros.network.HttpMessage;
+import org.zaproxy.zap.extension.ascan.VariantFactory;
+
+/**
+ * {@code AbstractAppVariantPlugin} is the abstract base class which is used to run per variant to
+ * modify multiple name value pairs of the {@code HttpMessage} per variant.
+ *
+ * @author KSASAN preetkaran20@gmail.com
+ */
+public abstract class AbstractAppVariantPlugin extends AbstractAppPlugin {
```

Comment:
@thc202 @kingthorin was planning to have another base class for usecases where scanrule wants to modify multiple name value pairs which is a use case for FileUpload scan rule. WDYT ?

Reply:
Wouldn’t file upload use the multipart variant which can already handle manipulation of various elements? (File name, disposition, content type)

Reply:
No no, Variant handles the manipulation but AbstractAppParamPlugin calls the scan method for one field at a time hence this new class gives the access to entire variant so that child can manipulate fields as per there need.

Reply:
Ah okay, I see what you're after.

Reply:
Why not change `AbstractAppParamPlugin` to just expose all parameters if that's what's needed?
e.g. `scan(HttpMessage, List<NameValuePair>)`
Those interested in all parameters just need to override the method.

Worth noting that variants do not support changing multiple parameters at the same time.

Reply:
@thc202 yeah that also can be done but i think it is ok to expose Variant because there are some extra methods/utilities present in variant which can be utilised  like setparameter/setEscapeParameter etc methods  else we need to expose them from parent.

Oh variants don't support changing multiple parameters ? because of start index and end index of value getting changed after setparameter is invoked ?
How should be proceed on this ? shall we re-initialise the variant with new message after setting paramter ?

Reply:
The `NameValuePair`s are already being exposed through `scan(HttpMessage, NameValuePair)` (and there are scan rules that make use of it), the only difference is that we would be exposing all of them at once. Those methods would not be irrelevant, they are used by scan rules that scan just one parameter at the same time.

The variants don't keep track of what changes were already done to a given message, the parameters would have to be set at the same time for properly construct the payload.

Reply:
no no, my point is that for the method that will expose list of name value pairs how are we going to call `setParameter/setEscapedParameter` methods?  we need to create 2 new methods which takes original nameValuePair as input and hence for someone who is implementing abstractAppParamPlugin it is tough to understand which method to invoke. I think new class creation is a clearly demarkation between the 2 usecases.

-----

```java
    private String setParameter(HttpMessage msg, NameValuePair originalPair, String value) {

        StringBuilder newBodyBuilder = new StringBuilder(msg.getRequestBody().toString());
        int idx = originalPair.getPosition() - 1;
        MultipartFormParameter mpPart = multiPartParams.get(idx);
        if (LOGGER.isDebugEnabled()) {
            LOGGER.debug(
                    "i: "
                            + idx
                            + " pos: "
                            + originalPair.getPosition()
                            + " S: "
                            + mpPart.getStart()
                            + " E: "
                            + mpPart.getEnd());
        }
        newBodyBuilder.replace(mpPart.getStart(), mpPart.getEnd(), value);

        String newBody = newBodyBuilder.toString();
        msg.getRequestBody().setBody(newBody);
        return newBody;
    }
```
What i was refering is that, the above method takes the start and end index from the already set HttpMessage in variant and if we set one value in the httpmessage, start and end index will not point to correct locations if we add another value to the httpmessage and hence it will break the httpmessage.

Reply:
I'm not sure it is, e.g. a list allows to add one element or multiple at the same time. If you have multiple parameters you call the method that accepts multiple, if you have one parameter you call the method that accepts one.

Yes, that's what I said in the previous comment, variants don't currently keep track of what changes were done to a given message, the parameters need to be set at the same time.

Reply:
So my point was that:
AbstractAppParamPlugin currently has following methods
```java
    protected String setParameter(HttpMessage message, String param, String value) {
        return variant.setParameter(message, originalPair, param, value);
    }

    protected String setEscapedParameter(HttpMessage message, String param, String value) {
        return variant.setEscapedParameter(message, originalPair, param, value);
    }
```

Now we need to add 2 new methods:
```java
 protected String setParameter(HttpMessage message, String param, String value) {
        return variant.setParameter(message, originalPair, param, value);
    }

    protected String setEscapedParameter(HttpMessage message, String param, String value) {
        return variant.setEscapedParameter(message, originalPair, param, value);
    }
 protected String setParameter(HttpMessage message, String param, String value, NameValuePair nameValuePair) {
        return variant.setParameter(message, nameValuePair, param, value);
    }

    protected String setEscapedParameter(HttpMessage message, String param, String value, NameValuePair nameValuePair) {
        return variant.setEscapedParameter(message, nameValuePair, param, value);
    }
```

I was thinking there is some difficulty in understanding and implementation where if child has called the original methods then it will cause some issues.

why to cause all this confusion.

Reply:
The proposed methods do not support changing multiple parameters. Probably better to discuss how the variant will support changing multiple parameters instead, that's what actually will define how the plugin will expose that.

Reply:
> There's still no proposal on how to change the parameters at the same time [#6371 (comment)](https://github.com/zaproxy/zaproxy/pull/6371#discussion_r555697929).

@thc202  Now this PR handles changing multiple parameters. Please look at PR: https://github.com/zaproxy/zaproxy/pull/6371/files#diff-9dfdfe7aed90f3b927955bcfb81a1c9ae34e745c7b5b4e04d79b5b0130cfb12aR241